src/Query/QueryBuilder.php

```diff
@@ -1564,4 +1577,52 @@ public function __clone()
             $this->params[$name] = clone $param;
         }
     }
+
+    /**
+     * Enables caching of the results of this query, for given amount of seconds
+     * and optionally specified witch key to use for the cache entry.
+     *
+     * @param int|null    $lifetime How long the cache entry is valid, in seconds.
+     * @param string|null $cacheKey Key to use for the cache entry.
+     *
+     * @return $this
+     */
+    public function enableResultCache(int $lifetime, ?string $cacheKey = null): self
+    {
+        $this->resultCacheProfile = new QueryCacheProfile($lifetime, $cacheKey);
+
+        return $this;
+    }
+
+    /**
+     * Disables caching of the results of this query.
+     *
+     * @return $this
+     */
+    public function disableResultCache(): self
+    {
+        $this->resultCacheProfile = null;
+
+        return $this;
+    }
+
+    /**
+     * Gets the cache profile for the result caching.
+     */
+    public function getResultCacheProfile(): ?QueryCacheProfile
+    {
+        return $this->resultCacheProfile;
+    }
+
+    /**
+     * Set a cache profile for the result caching.
+     *
+     * @return $this
+     */
+    public function setResultCacheProfile(?QueryCacheProfile $cacheProfile): self
```

Comment:
Having two ways to enable-disable caching looks like overkill. When a method just "sets" something, it looks creepy. The name of the method should convey some semantics.

How about having `enableResultCache()` accept an instance of `QueryCacheProfile`? That should be sufficient.

Reply:
I took inspiration from `doctrine/orm`, where something similar is implemented. There are a lot of methods for result caching: `setResultCacheProfile`, `setResultCache`,  `useResultCache`, `enableResultCache`, `disableResultCache`, `setResultCacheLifetime`,  `expireResultCache`, `getExpireResultCache`, `getQueryCacheProfile`, `setResultCacheId`, `setResultCacheDriver` (deprecated), `getResultCacheDriver` (deprecated), `getResultCacheLifetime` (deprecated), `getResultCacheId` (deprecated) in `AbstractQuery` class. Too much methods for the same thing. I also don't like many methods for the same thing and that's why I chose only 4. The `enableResultCache` where you can simply pass the lifetime and optionally cache key and it's not necessary create new instance of `QueryCacheProfile` every time and if someone need more control, then he can use `setQueryCacheProfile`. But I think that for many cases setting the lifetime will be sufficient. And I also wanted to keep the naming and functioning of the methods according to `doctrine/orm` package. If someone is used to using these methods in `doctrine/orm', it would be good to keep their functionality similar. 

https://github.com/doctrine/orm/blob/2.12.x/lib/Doctrine/ORM/AbstractQuery.php#L717

```php
// variant A
$builder->enableResultCache(300)->fetchAllAssociative();

// variant B
$builder->enableResultCache(new QueryCacheProfile(300))->fetchAllAssociative();
```

I like the variant A because of less typing and it is simillar with the method in `doctrine/orm`. What do you thing about it?

Reply:
The documentation suggests that the `QueryCacheProfile` instance needs to be created by the user and passed to the connection:
https://github.com/doctrine/dbal/blob/3ae5312bbd0b5fe271e90ca8d957038d14e1c840/docs/en/reference/caching.rst#L42-L43

The `QueryCacheProfile` class doesn't have mandatory parameters in the constructors and have some builder-like methods:
https://github.com/doctrine/dbal/blob/595e2df1590257bc5bb0423a1ba1f332e01be0f3/src/Cache/QueryCacheProfile.php#L28-L36 https://github.com/doctrine/dbal/blob/595e2df1590257bc5bb0423a1ba1f332e01be0f3/src/Cache/QueryCacheProfile.php#L90-L103

With that said, I believe it's the user who should be responsible for building a `QueryCacheProfile` object and passing it to the query builder. I don't think we need to introduce another way of building `QueryCacheProfile`. If there are any changes in the `QueryCacheProfile` constructor, they would have to be reflected in the currently proposed `enableResultCache()`. I believe we should avoid that.

Reply:
I understand, I changed the code and I put some new tests there, can you please check it if it is enough thanks